File: web/src/lib/voice/voice-session.ts
/**
 * VoiceSession — the framework-free state machine that drives one voice
 * conversation with xAI's Grok Voice Agent over a direct WebSocket.
 *
 * It is DEPENDENCY-INJECTED: every side effect (socket, audio IO, HTTP to our
 * `/api/voice/*` routes, the clock) arrives through {@link VoiceSessionSeams},
 * so jsdom tests drive it with scripted fakes and never touch real WebSocket /
 * AudioContext / getUserMedia. The real DOM seams live in
 * `@/lib/voice/audio-io` and are assembled by the overlay.
 *
 * Lifecycle (plan §5, Workstream F):
 *   idle → connecting → listening ⇄ thinking ⇄ speaking → … → ended | error
 *
 * Responsibilities: fetch an ephemeral token, open the socket with the
 * `xai-client-secret.<token>` subprotocol, push a single `session.update` from
 * the bootstrap, stream mic audio, drive phase transitions, flush playback on
 * barge-in BEFORE any bookkeeping, batch parallel tool calls (all outputs then
 * one `response.create`), persist each finished turn fire-and-forget, answer
 * pings, and auto-end at the session cap. No React, no DOM here.
 */

import type {
  VoiceTokenRequestBody,
  VoiceToolRequestBody,
  VoiceTranscriptRequestBody,
  VoiceSessionBootstrap,
} from "@/lib/voice/voice-types";
import type { Format } from "@/data/formats";
import {
  clientSecretSubprotocol,
  parseServerEvent,
  realtimeUrl,
  voiceToolLabel,
  VOICE_TRANSCRIBE_MODEL,
  type ClientEvent,
  type ServerEvent,
} from "@/lib/voice/voice-protocol";

// ---------------------------------------------------------------------------
// Seams (injected)
// ---------------------------------------------------------------------------

/** A thin WebSocket wrapper — the only socket surface the machine touches. */
export interface VoiceSocket {
  send(data: string): void;
  close(): void;
  onOpen(cb: () => void): void;
  onMessage(cb: (data: string) => void): void;
  onClose(cb: () => void): void;
  onError(cb: (err: unknown) => void): void;
}

/** Real-time audio capture + playback, at a single fixed {@link sampleRate}. */
export interface VoiceAudioIO {
  /** The context sample rate (an xAI-supported value). */
  readonly sampleRate: number;
  /** Begin mic capture; `onChunk` receives ~100 ms base64 PCM16 frames. */
  startCapture(onChunk: (base64Pcm16: string) => void): Promise<void>;
  /** Stop mic capture and release the microphone. */
  stopCapture(): void;
  /** Queue a base64 PCM16 playback chunk. */
  enqueue(base64Pcm16: string): void;
  /** Stop + clear all queued/playing audio (barge-in). */
  flush(): void;
  /** Release the audio context entirely. */
  close(): void;
}

export interface VoiceSessionSeams {
  connect(url: string, subprotocol: string): VoiceSocket;
  audio: VoiceAudioIO;
  fetchToken(
    body: VoiceTokenRequestBody,
  ): Promise<{ token: string; expires_at: number; session: VoiceSessionBootstrap }>;
  execTool(
    body: VoiceToolRequestBody,
  ): Promise<{ output: unknown }>;
  postTranscript(body: VoiceTranscriptRequestBody): Promise<void>;
  now(): number;
}

export interface VoiceSessionOptions {
  sessionId: string;
  format: Format;
  seams: VoiceSessionSeams;
}

// ---------------------------------------------------------------------------
// Observable state
// ---------------------------------------------------------------------------

export type VoicePhase =
  | "idle"
  | "connecting"
  | "listening"
  | "thinking"
  | "speaking"
  | "ended"
  | "error";

export interface VoiceSessionState {
  phase: VoicePhase;
  /** User-readable message when `phase === "error"`, else null. */
  error: string | null;
}

/** Live captions for the CURRENT turn (both reset when a turn completes). */
export interface VoiceCaption {
  user: string;
  assistant: string;
}

/** One tool the model invoked this turn, for the overlay ticker. */
export interface VoiceToolActivity {
  id: string;
  tool: string;
  label: string;
}

type StateListener = (state: VoiceSessionState) => void;
type CaptionListener = (caption: VoiceCaption) => void;
type ToolListener = (activities: VoiceToolActivity[]) => void;

export class VoiceSession {
  private readonly sessionId: string;
  private readonly format: Format;
  private readonly seams: VoiceSessionSeams;

  private socket: VoiceSocket | null = null;
  private phase: VoicePhase = "idle";
  private error: string | null = null;

  // Per-turn caption accumulators.
  private userText = "";
  private assistantText = "";
  private toolActivities: VoiceToolActivity[] = [];

  // Function calls collected during the in-flight model response. Kept until
  // `response.done` closes the batch (all parallel calls are in by then).
  private pendingCalls: { call_id: string; promise: Promise<{ output: unknown }> }[] =
    [];

  private maxSessionTimer: ReturnType<typeof setTimeout> | null = null;
  private startedAt = 0;
  private finished = false; // end() called or a terminal error/timeout reached

  private readonly stateListeners = new Set<StateListener>();
  private readonly captionListeners = new Set<CaptionListener>();
  private readonly toolListeners = new Set<ToolListener>();

  constructor(opts: VoiceSessionOptions) {
    this.sessionId = opts.sessionId;
    this.format = opts.format;
    this.seams = opts.seams;
  }

  // ── Subscriptions ────────────────────────────────────────────────────────

  onState(cb: StateListener): () => void {
    this.stateListeners.add(cb);
    cb(this.snapshot());
    return () => this.stateListeners.delete(cb);
  }

  onCaption(cb: CaptionListener): () => void {
    this.captionListeners.add(cb);
    return () => this.captionListeners.delete(cb);
  }

  onToolActivity(cb: ToolListener): () => void {
    this.toolListeners.add(cb);
    return () => this.toolListeners.delete(cb);
  }

  getPhase(): VoicePhase {
    return this.phase;
  }

  /** Milliseconds since the socket opened (0 before connect). */
  elapsedMs(): number {
    return this.startedAt === 0 ? 0 : this.seams.now() - this.startedAt;
  }

  // ── Lifecycle ──────────────────────────────────────────────────────────

  async start(): Promise<void> {
    if (this.phase !== "idle") return;
    this.setPhase("connecting");

    let bootstrap: VoiceSessionBootstrap;
    try {
      const res = await this.seams.fetchToken({
        session_id: this.sessionId,
        format: this.format,
      });
      bootstrap = res.session;
      this.connectSocket(res.token, bootstrap);
    } catch (err) {
      this.fail(errMessage(err, "Couldn't start voice mode."));
    }
  }

  end(): void {
    if (this.finished) {
      // Still ensure resources are released if end() races itself.
      this.teardown();
      return;
    }
    this.finished = true;
    this.teardown();
    if (this.phase !== "error") this.setPhase("ended");
  }

  // ── Socket wiring ────────────────────────────────────────────────────────

  private connectSocket(token: string, bootstrap: VoiceSessionBootstrap): void {
    let socket: VoiceSocket;
    try {
      socket = this.seams.connect(
        realtimeUrl(bootstrap.model),
        clientSecretSubprotocol(token),
      );
    } catch (err) {
      this.fail(errMessage(err, "Couldn't connect to the voice service."));
      return;
    }
    this.socket = socket;

    socket.onOpen(() => {
      if (this.finished) return;
      this.startedAt = this.seams.now();
      this.scheduleMaxSession(bootstrap.max_session_ms);
      this.send(this.buildSessionUpdate(bootstrap));
      void this.beginCapture();
    });
    socket.onMessage((data) => this.handleMessage(data));
    socket.onClose(() => {
      if (this.finished) return;
      // A clean close we didn't initiate — end the session gracefully.
      this.finished = true;
      this.teardown();
      if (this.phase !== "error") this.setPhase("ended");
    });
    socket.onError((err) => {
      if (this.finished) return;
      this.fail(errMessage(err, "The voice connection dropped."));
    });
  }

  private async beginCapture(): Promise<void> {
    try {
      await this.seams.audio.startCapture((chunk) => {
        if (this.finished) return;
        this.send({ type: "input_audio_buffer.append", audio: chunk });
      });
      if (!this.finished) this.setPhase("listening");
    } catch (err) {
      this.fail(errMessage(err, "Couldn't access the microphone."));
    }
  }

  private buildSessionUpdate(bootstrap: VoiceSessionBootstrap): ClientEvent {
    const rate = this.seams.audio.sampleRate;
    return {
      type: "session.update",
      session: {
        instructions: bootstrap.instructions,
        voice: bootstrap.voice,
        turn_detection: {
          type: "server_vad",
          idle_timeout_ms: bootstrap.idle_timeout_ms,
        },
        audio: {
          input: {
            format: { type: "audio/pcm", rate },
            transcription: { model: VOICE_TRANSCRIBE_MODEL },
          },
          output: { format: { type: "audio/pcm", rate } },
        },
        reasoning: { effort: bootstrap.reasoning_effort },
        tools: bootstrap.tools.map((t) => ({
          type: "function" as const,
          name: t.name,
          description: t.description,
          parameters: t.parameters,
        })),
      },
    };
  }

  // ── Server event handling ────────────────────────────────────────────────

  private handleMessage(data: string): void {
    if (this.finished) return;
    const event = parseServerEvent(data);
    if (event === null) return;
    switch (event.type) {
      case "input_audio_buffer.speech_started":
        // BARGE-IN: flush playback FIRST, before any state bookkeeping.
        this.seams.audio.flush();
        // A new user utterance starts a fresh turn — clear the ticker + the
        // interrupted assistant caption so the next answer renders clean.
        this.assistantText = "";
        this.toolActivities = [];
        this.emitCaption();
        this.emitTools();
        this.setPhase("listening");
        break;
      case "response.created":
        // A fresh model response — reset its assistant caption accumulator.
        this.assistantText = "";
        this.setPhase("thinking");
        break;
      case "response.output_audio.delta":
        this.seams.audio.enqueue(event.delta);
        this.setPhase("speaking");
        break;
      case "response.output_audio_transcript.delta":
        this.assistantText += event.delta;
        this.emitCaption();
        this.setPhase("speaking");
        break;
      case "conversation.item.input_audio_transcription.completed":
        this.userText = event.transcript;
        this.emitCaption();
        break;
      case "response.function_call_arguments.done":
        this.collectFunctionCall(event);
        break;
      case "response.done":
        this.handleResponseDone();
        break;
      case "ping":
        this.send({
          type: "pong",
          ping_timestamp: event.ping_timestamp ?? this.seams.now(),
        });
        break;
      case "error":
        this.handleServerError(event);
        break;
      // session.created / session.updated / speech_stopped / committed /
      // transcript.done — no state change needed.
      default:
        break;
    }
  }

  private collectFunctionCall(event: {
    name: string;
    call_id: string;
    arguments: string;
  }): void {
    // Surface the call in the ticker immediately (the model announces it aloud
    // too — the persona says a filler before calling).
    this.toolActivities = [
      ...this.toolActivities,
      { id: event.call_id, tool: event.name, label: voiceToolLabel(event.name) },
    ];
    this.emitTools();
    this.setPhase("thinking");
    // Execute eagerly (concurrently across parallel calls); the result is
    // gathered at the batch boundary (response.done).
    const promise = this.seams
      .execTool({
        session_id: this.sessionId,
        format: this.format,
        name: event.name,
        arguments: event.arguments,
      })
      .catch((err) => ({
        output: {
          error: "tool_failed",
          detail: errMessage(err, "tool request failed"),
        },
      }));
    this.pendingCalls.push({ call_id: event.call_id, promise });
  }

  private handleResponseDone(): void {
    if (this.pendingCalls.length > 0) {
      // This response was a batch of function calls — send every output, then
      // exactly ONE response.create so the model continues from the results.
      const batch = this.pendingCalls;
      this.pendingCalls = [];
      void Promise.all(
        batch.map(async ({ call_id, promise }) => ({
          call_id,
          output: (await promise).output,
        })),
      ).then((results) => {
        if (this.finished) return;
        for (const { call_id, output } of results) {
          this.send({
            type: "conversation.item.create",
            item: {
              type: "function_call_output",
              call_id,
              output: safeStringify(output),
            },
          });
        }
        this.send({ type: "response.create" });
      });
      return;
    }

    // A spoken turn finished. Persist the pair fire-and-forget, then reset the
    // per-turn accumulators and return to listening.
    if (this.userText.trim() && this.assistantText.trim()) {
      void this.seams
        .postTranscript({
          session_id: this.sessionId,
          format: this.format,
          user_text: this.userText,
          assistant_text: this.assistantText,
        })
        .catch((err) => {
          // A persistence failure must never break the live session.
          // eslint-disable-next-line no-console
          console.warn("voice transcript post failed", err);
        });
    }
    this.userText = "";
    this.assistantText = "";
    this.emitCaption();
    this.setPhase("listening");
  }

  private handleServerError(event: { code?: string; message?: string }): void {
    if (event.code === "timeout" || event.code === "max_duration") {
      // A benign end-of-session signal — tear down cleanly, not as an error.
      this.finished = true;
      this.teardown();
      this.setPhase("ended");
      return;
    }
    this.fail(event.message ?? "The voice service reported an error.");
  }

  // ── Timers + teardown ──────────────────────────────────────────────────

  private scheduleMaxSession(maxMs: number): void {
    if (!Number.isFinite(maxMs) || maxMs <= 0) return;
    this.maxSessionTimer = setTimeout(() => {
      if (this.finished) return;
      this.finished = true;
      this.teardown();
      this.setPhase("ended");
    }, maxMs);
  }

  private teardown(): void {
    if (this.maxSessionTimer !== null) {
      clearTimeout(this.maxSessionTimer);
      this.maxSessionTimer = null;
    }
    try {
      this.seams.audio.stopCapture();
      this.seams.audio.flush();
      this.seams.audio.close();
    } catch {
      /* audio already torn down — ignore */
    }
    try {
      this.socket?.close();
    } catch {
      /* socket already closed — ignore */
    }
    this.socket = null;
  }

  private fail(message: string): void {
    this.finished = true;
    this.teardown();
    this.error = message;
    this.setPhase("error");
  }

  // ── Emit helpers ─────────────────────────────────────────────────────────

  private send(event: ClientEvent): void {
    if (!this.socket) return;
    try {
      this.socket.send(JSON.stringify(event));
    } catch {
      /* a send on a closing socket is not fatal — the close/error path wins */
    }
  }

  private snapshot(): VoiceSessionState {
    return { phase: this.phase, error: this.error };
  }

  private setPhase(phase: VoicePhase): void {
    if (this.phase === phase) return;
    this.phase = phase;
    const snap = this.snapshot();
    for (const cb of this.stateListeners) cb(snap);
  }

  private emitCaption(): void {
    const caption: VoiceCaption = {
      user: this.userText,
      assistant: this.assistantText,
    };
    for (const cb of this.captionListeners) cb(caption);
  }

  private emitTools(): void {
    const list = this.toolActivities;
    for (const cb of this.toolListeners) cb(list);
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function errMessage(err: unknown, fallback: string): string {
  if (err instanceof Error && err.message) return err.message;
  if (typeof err === "string" && err) return err;
  return fallback;
}

/** Stringify a tool result; a circular/unserializable shape degrades to error. */
function safeStringify(value: unknown): string {
  try {
    return JSON.stringify(value ?? null);
  } catch {
    return JSON.stringify({ error: "unserializable_tool_output" });
  }
}

// Type used by ServerEvent narrowing above; re-export nothing (internal only).
export type { ServerEvent };

File: ios/OakApp/Features/Voice/VoiceSession.swift
import Foundation
import Observation

/// The heart of voice mode — the framework-free state machine that drives one
/// spoken conversation with xAI's Grok Voice Agent over a direct WebSocket. A
/// faithful port of `web/src/lib/voice/voice-session.ts`; the web source is
/// authoritative on every transition.
///
/// `@MainActor @Observable` (like ``ChatViewModel``): all state mutates on the
/// main actor and the overlay observes it directly. It is DEPENDENCY-INJECTED —
/// every side effect (the socket, audio IO, the `/api/voice/*` HTTP calls, the
/// clock) arrives through a protocol, so tests drive it with scripted fakes and
/// never touch a real `URLSessionWebSocketTask` / `AVAudioEngine` / the network.
///
/// Lifecycle: `idle → connecting → listening ⇄ thinking ⇄ speaking → … →
/// ended | error`. Responsibilities: mint an ephemeral token, open the socket
/// with the `xai-client-secret.<token>` subprotocol, push a single
/// `session.update`, stream mic audio, drive phase transitions, flush playback on
/// barge-in BEFORE any bookkeeping, batch parallel tool calls (all outputs then
/// exactly one `response.create`), persist each finished turn fire-and-forget,
/// answer pings, and auto-end at the session cap.
@MainActor
@Observable
final class VoiceSession {

  // MARK: Observable state

  /// The coarse conversation phase, driving the overlay's orb + status copy.
  private(set) var phase: VoicePhase = .idle

  /// A user-readable message when ``phase`` is ``VoicePhase/error``, else `nil`.
  private(set) var errorMessage: String?

  /// Live captions for the CURRENT turn (both fields reset when a turn completes).
  private(set) var caption = VoiceCaption(user: "", assistant: "")

  /// The tools the model invoked this turn, for the overlay ticker. Cleared on a
  /// fresh user utterance (barge-in).
  private(set) var toolActivities: [VoiceToolActivity] = []

  /// When the session started streaming (set once capture begins), for an
  /// elapsed-time readout. `nil` before connect.
  private(set) var startedAt: Date?

  // MARK: Dependencies + identity

  @ObservationIgnored private let service: any VoiceService
  @ObservationIgnored private let connect: @Sendable (URL, String) -> any VoiceRealtimeConnection
  @ObservationIgnored private let audio: any VoiceAudioIO
  @ObservationIgnored private let clock: any VoiceClock
  @ObservationIgnored private let sessionId: String
  @ObservationIgnored private let format: Format

  // MARK: Internal machine state (not observed)

  @ObservationIgnored private var connection: (any VoiceRealtimeConnection)?
  @ObservationIgnored private var inboundTask: Task<Void, Never>?
  @ObservationIgnored private var captureTask: Task<Void, Never>?
  @ObservationIgnored private var maxSessionTask: Task<Void, Never>?

  /// Function calls collected during the in-flight model response, kept until
  /// `response.done` closes the batch (all parallel calls are in by then). Each
  /// tool executes eagerly as its own task; the result is gathered at the batch
  /// boundary.
  @ObservationIgnored private var pendingCalls: [(callId: String, task: Task<JSONValue, Never>)] = []

  /// `end()` called, or a terminal error/timeout/close reached. Guards every
  /// await resumption so a torn-down session never mutates state or sends.
  @ObservationIgnored private var finished = false

  init(
    service: any VoiceService,
    connect: @escaping @Sendable (URL, String) -> any VoiceRealtimeConnection,
    audio: any VoiceAudioIO,
    clock: any VoiceClock,
    sessionId: String,
    format: Format
  ) {
    self.service = service
    self.connect = connect
    self.audio = audio
    self.clock = clock
    self.sessionId = sessionId
    self.format = format
  }

  // MARK: Lifecycle

  /// Mints a token, opens the socket, configures the session, and starts capture.
  /// A no-op unless the session is still idle.
  func start() async {
    guard phase == .idle else { return }
    phase = .connecting

    // 1. Mint the ephemeral token + session bootstrap.
    let bootstrap: VoiceTokenResponse
    do {
      bootstrap = try await service.fetchToken(sessionId: sessionId, format: format)
    } catch {
      fail(Self.startFailureMessage(error))
      return
    }
    guard !finished else { return }

    // 2. Open the socket with the token as the client-secret subprotocol.
    let conn = connect(
      VoiceRealtimeProtocol.realtimeURL(model: bootstrap.session.model),
      VoiceRealtimeProtocol.clientSecretSubprotocol(token: bootstrap.token)
    )
    connection = conn

    // 3. Consume inbound frames; an unexpected finish is a clean end (web onClose).
    inboundTask = Task { [weak self] in
      for await text in conn.inbound() {
        guard let self, !self.finished else { return }
        self.handle(parseServerEvent(text))
      }
      self?.end()
    }

    // 4. Push the single session.update built from the bootstrap + our sample rate.
    await conn.send(
      VoiceClientEvent.sessionUpdate(
        instructions: bootstrap.session.instructions,
        voice: bootstrap.session.voice,
        idleTimeoutMs: bootstrap.session.idleTimeoutMs,
        sampleRate: audio.sampleRate,
        reasoningEffort: bootstrap.session.reasoningEffort,
        tools: bootstrap.session.tools
      ).encode()
    )
    guard !finished else { return }

    // 5. Arm the client-side auto-end at the session cap.
    let maxMs = bootstrap.session.maxSessionMs
    if maxMs > 0 {
      maxSessionTask = Task { [weak self, clock] in
        try? await clock.sleep(forMilliseconds: maxMs)
        guard let self, !self.finished else { return }
        self.end()
      }
    }

    startedAt = clock.now()

    // 6. Start mic capture; forward each chunk as an input-audio append.
    let chunks: AsyncStream<String>
    do {
      chunks = try await audio.startCapture()
    } catch {
      fail("Microphone unavailable.")
      return
    }
    guard !finished else { return }
    captureTask = Task { [weak self] in
      for await chunk in chunks {
        guard let self, !self.finished else { return }
        await self.connection?.send(VoiceClientEvent.inputAudioAppend(base64: chunk).encode())
      }
    }

    phase = .listening
  }

  /// Ends the session and releases every resource. Idempotent — a second call
  /// (or a race with a terminal error/timeout/close) is a no-op, so teardown runs
  /// exactly once.
  func end() {
    guard !finished else { return }
    finished = true
    teardown()
    if phase != .error { phase = .ended }
  }

  // MARK: Server event handling

  /// Folds one parsed server event into the machine. `nil` (malformed / unknown
  /// frame) and the ignorable session/lifecycle events are no-ops. Every
  /// transition mirrors `handleMessage` in the web source.
  private func handle(_ event: VoiceServerEvent?) {
    guard !finished, let event else { return }
    switch event {
    case .speechStarted:
      // BARGE-IN: flush playback FIRST, before any state bookkeeping. A new user
      // utterance starts a fresh turn — clear the ticker + the interrupted
      // assistant caption so the next answer renders clean.
      audio.flush()
      caption.assistant = ""
      toolActivities = []
      phase = .listening

    case .responseCreated:
      caption.assistant = ""
      phase = .thinking

    case let .audioDelta(base64):
      audio.enqueue(base64)
      phase = .speaking

    case let .transcriptDelta(text):
      caption.assistant += text
      phase = .speaking

    case .transcriptDone:
      break

    case let .userTranscript(text):
      caption.user = text

    case let .functionCallDone(name, callId, arguments):
      collectFunctionCall(name: name, callId: callId, arguments: arguments)

    case .responseDone:
      handleResponseDone()

    case let .ping(timestamp):
      let ts = timestamp ?? clock.now().timeIntervalSince1970 * 1000
      Task { [weak self] in
        await self?.connection?.send(VoiceClientEvent.pong(pingTimestamp: ts).encode())
      }

    case let .errorEvent(code, message):
      if code == "timeout" || code == "max_duration" {
        // A benign end-of-session signal — tear down cleanly, not as an error.
        end()
      } else {
        fail(message ?? "Voice connection failed.")
      }

    case .sessionCreated, .sessionUpdated, .speechStopped, .committed:
      break
    }
  }

  private func collectFunctionCall(name: String, callId: String, arguments: String) {
    // Surface the call in the ticker immediately (the persona also says a filler
    // before it calls, so this is visible in step with the speech).
    toolActivities.append(VoiceToolActivity(id: callId, name: name, label: voiceToolLabel(name)))
    phase = .thinking
    // Execute eagerly (concurrently across parallel calls); the result is
    // gathered at the batch boundary (response.done). Capture the seam + ids by
    // value so the task holds no reference back to the session.
    let call = Task { [service, sessionId, format] () -> JSONValue in
      do {
        return try await service.execTool(sessionId: sessionId, format: format, name: name, arguments: arguments)
      } catch {
        return Self.toolFailure(error)
      }
    }
    pendingCalls.append((callId: callId, task: call))
  }

  private func handleResponseDone() {
    if !pendingCalls.isEmpty {
      // This response was a batch of function calls — send every output IN ORDER,
      // then exactly ONE response.create so the model continues from the results.
      let batch = pendingCalls
      pendingCalls = []
      Task { [weak self] in
        for (callId, task) in batch {
          let output = await task.value
          guard let self, !self.finished else { return }
          await self.connection?.send(
            VoiceClientEvent.functionCallOutput(callId: callId, output: output.serialized()).encode()
          )
        }
        guard let self, !self.finished else { return }
        await self.connection?.send(VoiceClientEvent.responseCreate.encode())
      }
      return
    }

    // A spoken turn finished. Persist the pair fire-and-forget (never awaited on
    // the live path), then reset the per-turn accumulators and return to listening.
    let hasUser = !caption.user.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    let hasAssistant = !caption.assistant.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    if hasUser && hasAssistant {
      let userText = caption.user
      let assistantText = caption.assistant
      Task { [service, sessionId, format] in
        await service.postTranscript(
          sessionId: sessionId,
          format: format,
          userText: userText,
          assistantText: assistantText
        )
      }
    }
    caption = VoiceCaption(user: "", assistant: "")
    phase = .listening
  }

  // MARK: Teardown

  /// Cancels every task, releases audio + the socket. Called exactly once, from
  /// ``end()`` or ``fail(_:)``. Mirrors the web teardown order (timer, audio, then
  /// socket).
  private func teardown() {
    maxSessionTask?.cancel()
    maxSessionTask = nil
    inboundTask?.cancel()
    inboundTask = nil
    captureTask?.cancel()
    captureTask = nil
    audio.stopCapture()
    audio.flush()
    audio.close()
    connection?.close()
    connection = nil
  }

  private func fail(_ message: String) {
    guard !finished else { return }
    finished = true
    teardown()
    errorMessage = message
    phase = .error
  }

  // MARK: Helpers

  /// Maps a token-mint failure to user-facing copy. An auth failure (voice is
  /// signed-in only) points the user at signing in; everything else is generic.
  private static func startFailureMessage(_ error: Error) -> String {
    if let oak = error as? OakError, case .unauthorized = oak {
      return "Sign in to use voice mode."
    }
    return "Couldn't start voice mode."
  }

  /// Folds a thrown tool call into the error object the web sends back as the
  /// function's output (`{ error: "tool_failed", detail: <message> }`), so the
  /// model can recover instead of the socket stalling on a missing output.
  static func toolFailure(_ error: Error) -> JSONValue {
    .object(["error": .string("tool_failed"), "detail": .string(toolFailureDetail(error))])
  }

  private static func toolFailureDetail(_ error: Error) -> String {
    if let oak = error as? OakError {
      switch oak {
      case let .http(_, _, message) where !message.isEmpty: return message
      case let .transport(underlying): return underlying
      default: break
      }
    }
    return "tool request failed"
  }
}

// MARK: - Observable value types

/// The coarse phase of a voice conversation, driving the overlay orb + status.
enum VoicePhase: Equatable, Sendable {
  case idle
  case connecting
  case listening
  case thinking
  case speaking
  case ended
  case error
}

/// Live captions for the current turn — both reset when the turn completes.
struct VoiceCaption: Equatable, Sendable {
  var user: String
  var assistant: String
}

/// One tool the model invoked this turn, for the overlay ticker. `id` is the
/// call id (stable + unique per call).
struct VoiceToolActivity: Identifiable, Equatable, Sendable {
  let id: String
  let name: String
  let label: String
}
